## Background

I got reproducible failures when I tried to build gettext[tools] on osx with port libxml2 installed. The error pointed to vcpkg's libxml2 where I would have expected to find usage of gettext's embedded libxml2. From running:
~~~
gcc -DHAVE_CONFIG_H -I. -I../.././../src/0.21-508c311661/libtextstyle/lib -I..  -I. -I../.././../src/0.21-508c311661/libtextstyle/lib -I.. -I../.././../src/0.21-508c311661/libtextstyle -Iglib -DIN_LIBTEXTSTYLE -DLIBXML_STATIC    -I../.././../src/0.21-508c311661/libtextstyle/lib/libcroco  -DDEPENDS_ON_LIBICONV=1   -fPIC -g -c ../.././../src/0.21-508c311661/libtextstyle/lib/libxml/globals.c
~~~
(no reference to vcpkg!) the errors were like:
~~~
../.././../src/0.21-508c311661/libtextstyle/lib/libxml/globals.c:367:7: error: illegal initializer (only variables can be initialized)
void *xmlGenericErrorContext = NULL;
      ^
/Users/me/dev/vcpkg/installed/x64-osx/include/libxml/globals.h:360:4: note: expanded from macro 'xmlGenericErrorContext'
(*(__xmlGenericErrorContext()))
   ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
~~~
I looked into vcpkg_configure_cmake and found:
~~~
# Used by GCC
set(ENV{C_INCLUDE_PATH} "${_VCPKG_INSTALLED}/include${VCPKG_HOST_PATH_SEPARATOR}${C_INCLUDE_PATH_BACKUP}")
~~~
Note that this creates values like:
~~~
/Users/me/dev/vcpkg/installed/x64-osx/include:
~~~
... including the trailing `:`, i.e. a separator followed by an empty element. [Gcc documentation states](https://gcc.gnu.org/onlinedocs/cpp/Environment-Variables.html#index-environment-variables) that the empty element is treated as the current directory, aka `-I.`

Now clang detects duplicate `.` elements:
~~~
clang -cc1 version 12.0.5 (clang-1205.0.22.9) default target x86_64-apple-darwin20.4.0
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/local/include"
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/Library/Frameworks"
ignoring duplicate directory "."
ignoring duplicate directory "../.././../src/0.21-508c311661/libtextstyle/lib"
ignoring duplicate directory ".."
ignoring duplicate directory "."
  as it is a non-system directory that duplicates a system directory
#include "..." search starts here:
#include <...> search starts here:
 ../.././../src/0.21-508c311661/libtextstyle/lib
 ..
 ../.././../src/0.21-508c311661/libtextstyle
 glib
 ../.././../src/0.21-508c311661/libtextstyle/lib/libcroco
 /usr/local/include
 /Users/kpa/dev/vcpkg/installed/x64-osx/include
 .
 /Library/Developer/CommandLineTools/usr/lib/clang/12.0.5/include
 /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include
 /Library/Developer/CommandLineTools/usr/include
 /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks (framework directory)
End of search list.
~~~
... and leaves it only after all explicit entries and the vcpkg include dir.

So that is why `libxml/globals.h` is not taken from `.` but from `/Users/me/dev/vcpkg/installed/x64-osx/include` with libxml2 installed.
And that's probably also the reason for #17892 where the earlier definition is from `/usr/local/include`.

--- a/scripts/cmake/vcpkg_configure_make.cmake
+++ b/scripts/cmake/vcpkg_configure_make.cmake
@@ -514,11 +514,18 @@ function(vcpkg_configure_make)
     endif()
     
     # Setup include environment (since these are buildtype independent restoring them is unnecessary)
+    macro(prepend_include_path var)
+        if("${${var}_BACKUP}" STREQUAL "")
+            set(ENV{${var}} "${_VCPKG_INSTALLED}/include")
+        else()
+            set(ENV{${var}} "${_VCPKG_INSTALLED}/include${VCPKG_HOST_PATH_SEPARATOR}${${var}_BACKUP}")
+        endif()
+    endmacro()
     # Used by CL 
-    set(ENV{INCLUDE} "${_VCPKG_INSTALLED}/include${VCPKG_HOST_PATH_SEPARATOR}${INCLUDE_BACKUP}")
+    prepend_include_path(INCLUDE)
     # Used by GCC
-    set(ENV{C_INCLUDE_PATH} "${_VCPKG_INSTALLED}/include${VCPKG_HOST_PATH_SEPARATOR}${C_INCLUDE_PATH_BACKUP}")
-    set(ENV{CPLUS_INCLUDE_PATH} "${_VCPKG_INSTALLED}/include${VCPKG_HOST_PATH_SEPARATOR}${CPLUS_INCLUDE_PATH_BACKUP}")
+    prepend_include_path(C_INCLUDE_PATH)
+    prepend_include_path(CPLUS_INCLUDE_PATH)
 
     # Flags should be set in the toolchain instead (Setting this up correctly requires a function named vcpkg_determined_cmake_compiler_flags which can also be used to setup CC and CXX etc.)
     if(VCPKG_TARGET_IS_WINDOWS)

